### ISSUE TITLE
Unicorn2 ARM64 PAC instructions cause EXCP_UDEF

### ISSUE THREAD

**Issue by blacktop** [unicorn-engine/unicorn#1522]
Hello!

I'm using the Golang bindings to emulate some iOS code and am hitting that interrupt every time I try and emulate a PAC instruction (i.e. `retab`). 

I do see that you have helpers like `pauth_addpac`, however, they are not exposed to the Go bindings.  Are those helpers functional now and if so can we add them to the bindings?

Thanks! 🙏 ❤️

**Comment by wtdcode** [comment]
I assume PAC support needs hw support which Unicorn can't help emulate. Is that correct?

**Comment by blacktop** [comment]
@wtdcode I also got a interrupt number 1 when trying to emulate a `casal` instruction. Does 1 mean unsupported instruction?

**Comment by blacktop** [comment]
Found this `#define EXCP_UDEF            1   /* undefined instruction */`

**Comment by blacktop** [comment]
I believe qemu can emulate PAC instructions using those helpers I mentioned above `pauth_addpac` etc and you would just fake it out by allowing me to add PAC when I load a MachO or dyld_shared_cache AND I can "apply" PAC too the pointers that the slide-info or chained-fixups call out and then the PAC instructions could fakeout "verifying" the signatures etc etc

**Comment by blacktop** [comment]
I built from the `dev` branch so I think that ARMv8.5-a isn't supported yet? #1324

**Comment by blacktop** [comment]
I going through your src now and it looks like you have everything you need to support PAC, is there a way for me to set what version of ARM64 I want to emulate?  I see in the Go bindings `SetCPUModel(model int) error` but don't see a list of availible models yet.

**Comment by blacktop** [comment]
`SetCPUModel` doesn't seem to have any effect and I've tried all of these:
```
	CPU_AARCH64_A57 = 0
	CPU_AARCH64_A53 = 1
	CPU_AARCH64_A72 = 2
	CPU_AARCH64_MAX = 3
```

**Comment by wtdcode** [comment]
Sorry for my late reply. If changing model doesn't work, I think it's a bug related to #1324

**Comment by wtdcode** [comment]
However, I need a PoC and perhaps a concrete context to fix this like #1520

**Comment by blacktop** [comment]
![dyld_emu](https://user-images.githubusercontent.com/6118188/148117606-aa237bc1-fff0-4a9d-9056-05a97768ddf3.png)

does this work as a POC? or do you mean you need a python script?

**Comment by blacktop** [comment]
But do see that qemu/you have models for ALL the PAC and ARMv8.2 `casal` instructions etc.  

So it really seems like you just need to expose them via some new ARM64 CPU models?

**Comment by blacktop** [comment]
Is the code in the `qemu` folder straight from QEMU? does that mean we just need to "re-sync" with QEMU to pull that in?

**Comment by wtdcode** [comment]
You have to post a minimal reproduction code to help me identify the bug. See the test code in #1520.

**Comment by blacktop** [comment]
So I also see that when it tries to parse the PAC instruction `xpaci` here:
```cpp
    case MAP(1, 0x01, 0x10): /* XPACI */
        if (!dc_isar_feature(aa64_pauth, s) || rn != 31) {
            goto do_unallocated;
        } else if (s->pauth_active) {
            tcg_rd = cpu_reg(s, rd);
            gen_helper_xpaci(tcg_ctx, tcg_rd, tcg_ctx->cpu_env, tcg_rd);
        }
        break;
```
It appears to fail the `dc_isar_feature(aa64_pauth, s)` check and then        
```cpp
void unallocated_encoding(DisasContext *s)
{
    /* Unallocated and reserved encodings are uncategorized */
    gen_exception_insn(s, s->pc_curr, EXCP_UDEF, syn_uncategorized(),
                       default_exception_el(s));
}
```
Which is exactly what I'm seeing.

However, why doesn't it die on the `pacibsp` ?
```cpp
    case 0x1b: // 0b11011: /* PACIBSP */
        if (s->pauth_active) {
            gen_helper_pacib(tcg_ctx, tcg_ctx->cpu_X[30], tcg_ctx->cpu_env, tcg_ctx->cpu_X[30], tcg_ctx->cpu_X[31]);
        }
        break;
```

I guess because it's a HINT it can be safely ignored on systems/CPU models that don't support PAC?

**Comment by blacktop** [comment]
Oh it looks like that `pauth_active` flag is actually set in `aarch64_tr_init_disas_context` here:
```cpp
    dc->pauth_active = FIELD_EX32(tb_flags, TBFLAG_A64, PAUTH_ACTIVE);
```    
I'm having trouble seeing where the `tb_flags` get set however.

**Comment by blacktop** [comment]
So I DO see for `CPU_AARCH64_MAX ` in `static void aarch64_max_initfn(struct uc_struct *uc, CPUState *obj)`:
That `casal` should be enabled here:
```cpp
    FIELD_DP64(t, ID_AA64ISAR0, ATOMIC, 2, t);
```
And PAC should be enabled here:
```cpp
    FIELD_DP64(t, ID_AA64ISAR1, APA, 1, t); /* PAuth, architected only */
    FIELD_DP64(t, ID_AA64ISAR1, API, 0, t);
    FIELD_DP64(t, ID_AA64ISAR1, GPA, 1, t);
    FIELD_DP64(t, ID_AA64ISAR1, GPI, 0, t);
```

**Comment by blacktop** [comment]
```go
package main

import (
	"encoding/binary"
	"fmt"
	"log"

	uc "github.com/unicorn-engine/unicorn/bindings/go/unicorn"
)

const (
	START_ADDR                 = 0x10000
	DATA_ADDR                  = START_ADDR + 0x200
	EXCP_UNDEFINED_INSTRUCTION = 1
	POISON                     = 0xdeadbeafdeadbeaf
)

func getCode() []byte {
	var code []byte
	instructions := []uint32{
		0xc8eafd28, // casal    x10, x8, [x9]
	}
	a := make([]byte, 4)
	for _, i := range instructions {
		binary.LittleEndian.PutUint32(a, i)
		code = append(code, a...)
	}
	return code
}

func main() {
	// setup
	mu, err := uc.NewUnicorn(uc.ARCH_ARM64, uc.MODE_ARM)
	if err != nil {
		log.Fatalf("failed to create new unicorn instance: %v", err)
	}
	// NOTE: using MAX which should set ALL FLAGS
	if err := mu.SetCPUModel(uc.CPU_AARCH64_MAX); err != nil {
		log.Fatalf("failed to set cpu model to CPU_AARCH64_MAX: %v", err)
	}

	// memmap
	if err := mu.MemMap(START_ADDR, 0x1000); err != nil {
		log.Fatalf("failed to memmap a page @ %#x: %v", START_ADDR, err)
	}

	// add CODE
	code := getCode()
	if err := mu.MemWrite(START_ADDR, code); err != nil {
		log.Fatalf("failed to write code at %#x: %v", START_ADDR, err)
	}

	// init state
	data := make([]byte, 8)
	if err := mu.MemWrite(DATA_ADDR, data); err != nil {
		log.Fatalf("failed to write 0 at %#x: %v", DATA_ADDR, err)
	}
	if err := mu.RegWrite(uc.ARM64_REG_X9, DATA_ADDR); err != nil {
		log.Fatalf("failed to write data addr to x9 reg: %v", err)
	}
	if err := mu.RegWrite(uc.ARM64_REG_X10, 0); err != nil {
		log.Fatalf("failed to write x10: %v", err)
	}
	if err := mu.RegWrite(uc.ARM64_REG_X8, POISON); err != nil {
		log.Fatalf("failed to write x8: %v", err)
	}

	// add HOOK_INTR
	if _, err := mu.HookAdd(uc.HOOK_INTR, func(mu uc.Unicorn, intno uint32) {
		switch intno {
		case EXCP_UNDEFINED_INSTRUCTION:
			log.Fatal("EXCP_UNDEFINED_INSTRUCTION")
		}
	}, 1, 0); err != nil {
		log.Fatalf("failed to register interrupt hook: %v", err)
	}

	// RUN
	if err := mu.Start(START_ADDR, START_ADDR+uint64(len(code))); err != nil {
		log.Fatalf("failed to emulate: %v", err)
	}

	// get results
	x9, err := mu.RegRead(uc.ARM64_REG_X9)
	if err != nil {
		log.Fatalf("failed to read x9: %v", err)
	}
	fmt.Printf("x9: %#08x\n", x9)
	x9Data, err := mu.MemRead(DATA_ADDR, 8)
	if err != nil {
		log.Fatalf("failed to read x9: %v", err)
	}
	fmt.Printf("[x9]: %#v\n", x9Data)

	x8, err := mu.RegRead(uc.ARM64_REG_X8)
	if err != nil {
		log.Fatalf("failed to read x9: %v", err)
	}
	fmt.Printf("x8: %#08x\n", x8)

	x10, err := mu.RegRead(uc.ARM64_REG_X10)
	if err != nil {
		log.Fatalf("failed to read x10: %v", err)
	}
	fmt.Printf("x10: %#08x\n", x10)
}
```

Ok correct me if I'm wrong @saagarjha 😏  but I believe this SHOULD:

1. compare x10 (which should be 0) to [x9] which should be 0
2. since the comp is equal it writes the val of x8 to memory
3. [x9] should now be `0xdeadbeafdeadbeaf`

However, what is DOES do is:
```bash
❯ CGO_ENABLED=1 CGO_LDFLAGS='-L/opt/homebrew/lib/ -Wl,-rpath -Wl,./unicorn2/build' go run emu.go
2022/01/04 16:01:45 EXCP_UNDEFINED_INSTRUCTION
exit status 1
```
😭

**Comment by blacktop** [comment]
@wtdcode is it possible that `mu.SetCPUModel(uc.CPU_AARCH64_MAX)` is having no effect?

**Comment by blacktop** [comment]
Also FYI the way I built unicorn2 was:
```bash
git clone -b dev https://github.com/unicorn-engine/unicorn.git
mkdir build
cd build
cmake  .. -DUNICORN_ARCH=aarch64 -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH=/opt/homebrew
make
make install
```

**Comment by wtdcode** [comment]
Fixed in c3a49766d8fe2530ce4c23cb3b0aed6e9be09ef2

**Comment by blacktop** [comment]
Oh WOW that was fast! Thank you @wtdcode ❤️
